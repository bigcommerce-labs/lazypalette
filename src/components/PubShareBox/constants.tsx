export enum Modes {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
    PREVIEW = 'PREVIEW',
    UNKNOWN = 'UNKNOWN',
}

export const PublishModalText = {
    [Modes.ACTIVE]: {
        action: 'Publish',
        body: 'Publishing your store will make all changes publicly visible',
        title: 'Publish Changes',
    },
    [Modes.INACTIVE]: {
        action: 'Use As Active Theme',
        body: 'You are about to make this the active theme in your store.',
        title: 'Use As Active Theme',
    },
};

export const ResetModalText = {
    action: 'Undo Changes',
    body: 'This will discard any changes made since the last time you saved.',
    title: 'Undo Changes',
};

export const purchaseEndPoint = (variationId: string) =>
    `/manage/marketplace/themes/${variationId}?action=purchase`;
