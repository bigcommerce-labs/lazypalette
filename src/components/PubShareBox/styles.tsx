import styled from 'styled-components';

interface ContainerProps {
    isChanged: boolean;
}

export const Container = styled.div.attrs<ContainerProps>({})`
    margin-left: auto;
    display: flex;
    justify-content: flex-end;
    width: 25rem;
    > div {
        margin-left: 1rem;
    }
`;
