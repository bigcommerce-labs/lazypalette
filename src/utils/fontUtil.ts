export function parseFont(font: string) {
    const provider = font.split('_')[0];
    switch (provider) {
        case 'Google':
            return parseGoogleFont(font);
        default:
            return null;
    }
}

function parseGoogleFont(font: string) {
    const split = font.split('_');
    const family = split[1];
    let formattedFont = '';
    let weight = split[2];

    if (split.length === 2) {
        formattedFont += `${family}|`;
    } else if (split.length > 2) {
        weight = weight.split(',')[0];
        formattedFont += `${family}:${weight}|`;
    }

    return `//fonts.googleapis.com/css?family=${formattedFont}`;
}
