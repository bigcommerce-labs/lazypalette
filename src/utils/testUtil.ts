export const sel = (id: string) => `[data-test-id="${id}"]`;
