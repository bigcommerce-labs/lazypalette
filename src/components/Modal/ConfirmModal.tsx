import React, { PureComponent } from 'react';

import ButtonInput from '../ButtonInput/ButtonInput';

import { Confirm, ConfirmBody, ConfirmButtons, ConfirmFooter } from './styles';
import Modal from './Modal';

interface ConfirmModalProps extends Partial<{
    body: string;
    title: string;
    primaryAction(): void;
    secondaryAction(): void;
    overlayClose(): void;
}> {}

const CTAButtons = ({ primaryAction, secondaryAction }: ConfirmModalProps) => (
    <ConfirmButtons>
        <ButtonInput
            onClick={primaryAction}
            classType="primary"
            type="button"
        >
            Cancel
        </ButtonInput>
        <ButtonInput
            onClick={secondaryAction}
            type="button"
        >
            OK
        </ButtonInput>
    </ConfirmButtons>
);

class ConfirmModal extends PureComponent<ConfirmModalProps> {
    static defaultProps: ConfirmModalProps = {
        title: 'Please Confirm:',
    };

    render() {
        const {
            body,
            primaryAction,
            secondaryAction,
            overlayClose,
            title,
        } = this.props;

        return (
            <Modal
                onClose={primaryAction}
                overlayClose={overlayClose}
                title={title}
                isTransparent={false}
            >
                <Confirm>
                    {body &&
                        <ConfirmBody>{body}</ConfirmBody>}
                    <ConfirmFooter>
                        <CTAButtons
                            primaryAction={primaryAction}
                            secondaryAction={secondaryAction}
                        />
                    </ConfirmFooter>
                </Confirm>
            </Modal>
        );
    }
}

export default ConfirmModal;
