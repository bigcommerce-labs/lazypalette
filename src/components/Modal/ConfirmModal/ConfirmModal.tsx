import React, { PureComponent } from 'react';

import ButtonInput from '../../ButtonInput/ButtonInput';

import { ConfirmButton, ConfirmButtons, ModalBody, ModalFooter, ModalView } from '../styles';
import Modal from '../Modal';

import { Messages } from '../constants';

interface ConfirmModalProps extends Partial<{
    title: string;
    primaryActionText?: string;
    primaryAction(): void;
    secondaryAction(): void;
    overlayClose(): void;
}> {}

const CTAButtons = ({ primaryAction, primaryActionText = Messages.Ok, secondaryAction }: ConfirmModalProps) => (
    <ConfirmButtons>
        <ConfirmButton>
            <ButtonInput
                onClick={secondaryAction}
                type="button"
                testId="cancel"
            >
                {Messages.Cancel}
            </ButtonInput>
        </ConfirmButton>
        <ConfirmButton>
            <ButtonInput
                classType="primary"
                onClick={primaryAction}
                type="button"
                testId={primaryActionText.replace(/\s+/g, '-').toLowerCase()}
            >
                {primaryActionText}
            </ButtonInput>
        </ConfirmButton>
    </ConfirmButtons>
);

class ConfirmModal extends PureComponent<ConfirmModalProps> {
    static defaultProps: ConfirmModalProps = {
        title: Messages.ConfirmTitle,
    };

    render() {
        const {
            children,
            primaryAction,
            secondaryAction,
            primaryActionText = Messages.Ok,
            overlayClose,
            title,
        } = this.props;

        return (
            <Modal
                onClose={secondaryAction}
                overlayClose={overlayClose}
                title={title}
                isTransparent={false}
            >
                <ModalView>
                    <ModalBody>{children}</ModalBody>
                    <ModalFooter>
                        <CTAButtons
                            primaryAction={primaryAction}
                            primaryActionText={primaryActionText}
                            secondaryAction={secondaryAction}
                        />
                    </ModalFooter>
                </ModalView>
            </Modal>
        );
    }
}

export default ConfirmModal;
