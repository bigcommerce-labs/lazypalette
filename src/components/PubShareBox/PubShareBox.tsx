import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import { State } from '../../reducers/reducers';
import { ThemeVariations, ThemeVariationHistory } from '../../reducers/theme';
import {
    trackPublish,
    trackResetCancel,
    trackResetClick,
    trackResetConfirmation,
    trackResetModalClose,
    trackSave,
} from '../../services/analytics';

import ButtonInput from '../ButtonInput/ButtonInput';
import ConfirmModal from '../Modal/ConfirmModal/ConfirmModal';

import { Modes, PublishModalText, ResetModalText } from './constants';
import { ButtonWrapper, Container } from './styles';
import ActiveAction from './ActiveAction/ActiveAction';
import InactiveAction from './InactiveAction/InactiveAction';
import PreviewAction from './PreviewAction/PreviewAction';

interface PubShareBoxProps {
    isChanged: boolean;
    isCurrent: boolean;
    isPrelaunchStore: boolean;
    isPurchased: boolean;
    configurationId: string;
    price: number;
    variations: ThemeVariations;
    variationHistory: ThemeVariationHistory;
    variationId: string;
    onPublish(): void;
    onReset(): void;
    onSave(): void;
}

interface PubShareBoxState {
    canPublish: boolean;
    canSave: boolean;
    isResetOpen: boolean;
    isPublishOpen: boolean;
}

export class PubShareBox extends PureComponent<PubShareBoxProps, PubShareBoxState> {
    readonly state: PubShareBoxState = {
        canPublish: false,
        canSave: false,
        isPublishOpen: false,
        isResetOpen: false,
    };

    handleModalCancel = (type: string) => {
        if (type === 'reset') {
            trackResetCancel();
            this.setState({ isResetOpen: false });
        } else {
            this.setState({ isPublishOpen: false });
        }
    };

    handleModalOpen = (type: string) => {
        if (type === 'reset') {
            trackResetClick();
            this.setState({ isResetOpen: true });
        } else {
            this.setState({ isPublishOpen: true });
        }
    };

    overlayClose = () => {
        trackResetModalClose();
        this.setState({ isResetOpen: false });
    };

    handleSave = () => {
        trackSave();
        this.props.onSave();
    };

    handleReset = () => {
        trackResetConfirmation();
        this.setState({ isResetOpen: false }, () => {
            this.props.onReset();
        });
    };

    handlePublish = () => {
        trackPublish(this.props.configurationId);
        this.setState({ isPublishOpen: false}, () => {
            this.props.onPublish();
        });
    };

    getActions = (mode: Modes) => {
        const { isPrelaunchStore, price, variationId } = this.props;
        const { canPublish, canSave } = this.state;

        switch (mode) {
            case Modes.ACTIVE:
                return (
                    <ActiveAction
                        isPrelaunchStore={isPrelaunchStore}
                        canPublish={canPublish}
                        canSave={canSave}
                        handleSave={this.handleSave}
                        handlePublish={isPrelaunchStore ? this.handlePublish : () => this.handleModalOpen('publish')}
                    />
                );
            case Modes.INACTIVE:
                return (
                    <InactiveAction
                        isPrelaunchStore={isPrelaunchStore}
                        canPublish={canPublish}
                        canSave={canSave}
                        handleSave={this.handleSave}
                        handlePublish={() => this.handleModalOpen('publish')}
                    />
                );
            case Modes.PREVIEW:
                return (
                    <PreviewAction
                        price={price}
                        variationId={variationId}
                    />
                );
            default:
                return null;
        }
    };

    getMode() {
        const { isCurrent, isPurchased, price } = this.props;

        if (isPurchased) {
            if (isCurrent) {
                return Modes.ACTIVE;
            } else {
                return Modes.INACTIVE;
            }
        } else if (price !== undefined) {
            return Modes.PREVIEW;
        }

        return Modes.UNKNOWN;
    }

    componentDidUpdate() {
        const { variations, variationHistory, configurationId } = this.props;
        const currentVariation = variations.filter(variation => variation.isCurrent);
        let canPublish = false;
        let canSave = true;
        if (variationHistory) {
            variationHistory.forEach(variation => {
                if (variation.configurationId === configurationId) {
                    canSave = false;
                    if (variation.type === 'saved' || currentVariation.length <= 0) {
                        canPublish = true;
                    }
                }
            });
        }

        if (canSave !== this.state.canSave) {
            this.setState({canSave});
        }

        if (canPublish !== this.state.canPublish) {
            this.setState({canPublish});
        }
    }

    render() {
        const {
            isChanged,
        } = this.props;
        const {
            isPublishOpen,
            isResetOpen,
        } = this.state;

        const mode = this.getMode();

        const publishModalText = PublishModalText[mode];

        return (
            <Container isChanged={isChanged}>
                {isChanged &&
                    <ButtonWrapper>
                        <ButtonInput
                            border={false}
                            onClick={() => this.handleModalOpen('reset')}
                            type="button"
                            testId="undo-changes"
                        >
                            Undo Changes
                        </ButtonInput>
                    </ButtonWrapper>
                }
                {this.getActions(mode)}
                {isResetOpen &&
                    <ConfirmModal
                        body={ResetModalText.body}
                        primaryAction={this.handleReset}
                        primaryActionText={ResetModalText.action}
                        secondaryAction={() => this.handleModalCancel('reset')}
                        overlayClose={this.overlayClose}
                        title={ResetModalText.title}
                    />
                }

                {isPublishOpen &&
                  <ConfirmModal
                      body={publishModalText.body}
                      primaryAction={this.handlePublish}
                      primaryActionText={publishModalText.action}
                      secondaryAction={() => this.handleModalCancel('publish')}
                      overlayClose={this.overlayClose}
                      title={publishModalText.title}
                  />
                }
            </Container>
        );
    }
}

const mapStateToProps = ({ theme, merchant }: State) => ({
    configurationId: theme.configurationId,
    isChanged: theme.isChanged,
    isCurrent: merchant.isCurrent,
    isPrelaunchStore: merchant.isPrelaunchStore,
    isPurchased: theme.isPurchased,
    price: theme.price,
    variationHistory: theme.variationHistory,
    variationId: theme.variationId,
    variations: theme.variations,
});

export default connect(mapStateToProps)(PubShareBox);
