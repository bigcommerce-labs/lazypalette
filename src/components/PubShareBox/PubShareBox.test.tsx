import { mount, shallow } from 'enzyme';
import React from 'react';

import {sel} from '../../utils/testUtil';

import ConfirmModal from '../Modal/ConfirmModal/ConfirmModal';

import { PubShareBox } from './PubShareBox';

describe('PubShareBox', () => {

    const undoChangesButton = sel('undo-changes');

    const mockVariations = [
        {
            configurationId: '123',
            defaultConfigurationId: '234',
            id: '567',
            isCurrent: true,
            lastCommitId: '',
            screenshot: {
                largePreview: 'host://meows/123.jpg',
                largeThumb: 'host://meows/234.jpg',
                smallThumb: 'host://meows/345.jpg',
            },
            themeId: '8900',
            variationName: 'light',
        },
    ];

    const mockVariationHistory = [
        {
            configurationId: '123',
            displayVersion: '2.3.2-rc.1',
            downloadProcessed: false,
            downloadable: true,
            installable: true,
            revisionId: 'e61f1460-8c45-0136-d583-0242ac110010',
            themeId: '0c33c3c0-885b-0136-0e49-0242ac110013',
            themeName: 'Cornerstone (2)',
            timestamp: '2018-08-27T16:40:58Z',
            type: 'saved',
            variationId: '0c7b7230-885b-0136-0e49-0242ac110013',
            variationName: 'Bold',
            versionId: '0c5a3140-885b-0136-0e49-0242ac110013',
        },
        {
            configurationId: '111',
            displayVersion: '222',
            downloadProcessed: false,
            downloadable: true,
            installable: true,
            revisionId: '333',
            themeId: '444',
            themeName: '555',
            timestamp: '2018-08-27T16:40:58Z',
            type: 'published',
            variationId: '666',
            variationName: 'Bold',
            versionId: '777',
        },
    ];
    const mockHandler = jest.fn();

    const pubShareBoxElement = <PubShareBox
        isChanged={false}
        isCurrent={true}
        isPrelaunchStore={false}
        isPurchased={true}
        price={0}
        configurationId="123"
        variationHistory={mockVariationHistory}
        variationId="234"
        variations={mockVariations}
        onPublish={mockHandler}
        onReset={mockHandler}
        onSave={mockHandler}
    />;

    const pubBoxShallow = shallow(pubShareBoxElement);

    it('renders', () => {
        expect(pubBoxShallow).toMatchSnapshot();
    });

    const pubBoxMount = mount(pubShareBoxElement);

    describe('isPurchased', () => {
        describe('when isPurchased is true', () => {
            describe('when isCurrent is true', () => {
                pubBoxMount.setProps({isPurchased: true});
                pubBoxMount.setProps({isCurrent: true});

                it('should render ActiveAction', () => {
                    expect(pubBoxMount.find('ActiveAction').length).toBe(1);
                });
                it('should not render InActiveAction', () => {
                    expect(pubBoxMount.find('InActiveAction').length).toBe(0);
                });
            });
        });

        describe('when isPurchased is false', () => {
            describe('when price is undefined', () => {
                it('should not render PreviewAction', () => {
                    pubBoxMount.setProps({isPurchased: false, price: undefined});
                    expect(pubBoxMount.find('PreviewAction').length).toBe(0);
                });
            });

            describe('when price is >= 0', () => {
                it('should render PreviewAction', () => {
                    pubBoxMount.setProps({isPurchased: false, price: 100});
                    expect(pubBoxMount.find('PreviewAction').length).toBe(1);
                });
            });
        });

        describe('isChange', () => {
            describe('is true', () => {
                it('should display "Undo changes" button', () => {
                    pubBoxMount.setProps({isChanged: true});
                    expect(pubBoxMount.find(undoChangesButton).hostNodes().length).toBe(1);
                });

                describe('when "Undo changes" button is clicked', () => {
                    it('opens ConfirmModal', () => {
                        pubBoxMount.find(undoChangesButton).hostNodes().simulate('click');
                        expect(pubBoxMount.find(ConfirmModal).length).toBe(1);
                    });
                });
            });
            describe('is false', () => {
                it('should not display "Undo changes" button', () => {
                    pubBoxMount.setProps({isChanged: false});
                    expect(pubBoxMount.find(undoChangesButton).hostNodes().length).toBe(0);
                });
            });
        });
    });
});
