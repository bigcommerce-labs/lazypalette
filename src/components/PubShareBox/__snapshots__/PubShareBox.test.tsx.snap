// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PubShareBox when rendering when the theme has not been purchased renders correctly 1`] = `
<styled.div
  isChanged={false}
>
  <PreviewAction
    price={100}
    variationId="234"
  />
</styled.div>
`;

exports[`PubShareBox when rendering when this is a purchased theme when this is not the active theme renders correctly 1`] = `
<styled.div
  isChanged={false}
>
  <InactiveAction
    canPublish={true}
    canSave={false}
    handlePublish={[Function]}
    handleSave={[Function]}
    isPrelaunchStore={false}
  />
</styled.div>
`;

exports[`PubShareBox when rendering when this is a purchased theme when this is the active theme when this is a prelaunch store renders correctly 1`] = `
<styled.div
  isChanged={false}
>
  <ActiveAction
    canPublish={true}
    canSave={false}
    handlePublish={[Function]}
    handleSave={[Function]}
    isPrelaunchStore={true}
  />
</styled.div>
`;

exports[`PubShareBox when rendering when this is a purchased theme when this is the active theme when this is not a prelaunch store renders correctly 1`] = `
<styled.div
  isChanged={false}
>
  <ActiveAction
    canPublish={true}
    canSave={false}
    handlePublish={[Function]}
    handleSave={[Function]}
    isPrelaunchStore={false}
  />
</styled.div>
`;
